### PR TITLE
[SPARK-47977] DateTimeUtils.timestampDiff and DateTimeUtils.timestampAdd should not throw INTERNAL_ERROR exception

### PR DESCRIPTION
**Pull request by vitaliili-db**
### What changes were proposed in this pull request?

Convert `INTERNAL_ERROR` for `timestampAdd` and `timestampDiff` to error with class. Reusing `INVALID_PARAMETER_VALUE.DATETIME_UNIT` used when parsing expressions.

The change is needed since `timestampDiff` and `timestampAdd` expressions could be constructed without going through parser - e.g. PySpark creates `timestampDiff` through PythonSQLUtils.

### Why are the changes needed?

Correct classification of errors

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing unit tests updated with new error class

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by vitaliili-db** [comment]
@MaxGekk please take a look

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 1]
```diff
@@ -698,7 +698,7 @@ object DateTimeUtils extends SparkDateTimeUtils {
       }
     } catch {
       case _: scala.MatchError =>
-        throw SparkException.internalError(s"Got the unexpected unit '$unit'.")
+        throw QueryExecutionErrors.invalidDatetimeUnitError("timestampAdd", unit)
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:701]: nit: TIMESTAMPADD

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 2]
```diff
@@ -736,7 +736,7 @@ object DateTimeUtils extends SparkDateTimeUtils {
       val endLocalTs = getLocalDateTime(endTs, zoneId)
       timestampDiffMap(unitInUpperCase)(startLocalTs, endLocalTs)
     } else {
-      throw SparkException.internalError(s"Got the unexpected unit '$unit'.")
+      throw QueryExecutionErrors.invalidDatetimeUnitError("timestampDiff", unit)
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:739]: nit: TIMESTAMPDIFF

**Review by gengliangwang — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 1 above)
Comment (vitaliili-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 2 above)
Comment (vitaliili-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Done

**Review by sigmod — APPROVED** [review]
thanks @vitaliili-db !

**Comment by cloud-fan** [comment]
it has merge conflicts, @vitaliili-db can you fix it?

**Comment by vitaliili-db** [comment]
@cloud-fan done!

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Pull request closed without merging**